File: 0437-path-sum-iii/0437-path-sum-iii.java
class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        if(root==null){
            return 0;
        }
        int x = selfsum(root,targetSum);
        int y = pathSum(root.left,targetSum);
        int z = pathSum(root.right,targetSum);
        
        return x+y+z;
    }
    public int selfsum(TreeNode root,long ts){
        if(root==null)return 0;
        
        int ans=0;
        if(ts-root.val==0){
            ans++;
        }
        int l = selfsum(root.left,ts-root.val);
        int r = selfsum(root.right,ts-root.val);
        
        return ans+l+r;        
    }
}

File: 0437-path-sum-iii/0437-path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* root,int ts,long long prefix,unordered_map<long long,int> &memo){
        if(root==nullptr)return 0;
        
        int ans = 0;
        prefix += root->val;
        ans += memo[prefix-ts];
        memo[prefix]++;
        
        ans += solve(root->left,ts,prefix,memo);
        ans += solve(root->right,ts,prefix,memo);
        
        memo[prefix]--;
        
        return ans;
    }
    int pathSum(TreeNode* root, int targetSum) {
        if(root == NULL)return 0;
        unordered_map<long long,int> memo;
        memo[0]=1;
        return solve(root,targetSum,0,memo);
    }
};